clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -530,14 +607,32 @@ private Map<TopicPartition, Long> beginningOrEndOffset(Collection<TopicPartition
                 timestampToSearch,
                 false);
         Map<TopicPartition, OffsetAndTimestamp> offsetAndTimestampMap =
-                eventHandler.addAndGet(listOffsetsEvent, timeout);
+                eventHandler.addAndGet(listOffsetsEvent, time.timer(timeout));
         return offsetAndTimestampMap.entrySet().stream().collect(Collectors.toMap(e -> e.getKey(),
                 e -> e.getValue().offset()));
     }
 
     @Override
     public OptionalLong currentLag(TopicPartition topicPartition) {
-        throw new KafkaException("method not implemented");
+        final Long lag = subscriptions.partitionLag(topicPartition, isolationLevel);
+
+        // if the log end offset is not known and hence cannot return lag and there is
+        // no in-flight list offset requested yet,
+        // issue a list offset request for that partition so that next time
+        // we may get the answer; we do not need to wait for the return value
+        // since we would not try to poll the network client synchronously
+        if (lag == null) {
+            if (subscriptions.partitionEndOffset(topicPartition, isolationLevel) == null &&
+                    !subscriptions.partitionEndOffsetRequested(topicPartition)) {
+                log.info("Requesting the log end offset for {} in order to compute lag", topicPartition);
+                subscriptions.requestPartitionEndOffset(topicPartition);
+                beginningOrEndOffset(Collections.singleton(topicPartition), ListOffsetsRequest.LATEST_TIMESTAMP, Duration.ofMillis(0));
```

Comment:
Here we could use `endOffsets(Collections.singleton(topicPartition), Duration.ofMillis(0));` instead for better readability I think

Reply:
I had it that way simply because it's a copy-and-paste of what's in the `KafkaConsumer` implementation of `currentLag`. But I'll change it for clarity.

Reply:
Update `currentLag`.